### ISSUE TITLE
JSDoc : support @interface and @implements

### ISSUE THREAD

**Issue by cyrilletuzi** [Microsoft/TypeScript#16142]
**TypeScript Version:**  2.3.3

**Code**

In a JavaScript file, with `checkJs` option : 

```js
/**
 * @interface Movable
 */

/**
 * @function
 * @name Movable#move
 * @param {number} distance
 */

/**
 * @implements {Movable}
 */
class Vehicle {}
```
**Actual behavior:**

No error.

**Expected behavior:**

An error should be raised. It's important in `checkJs` mode as JavaScript doesn't have native interfaces.

**Comment by pulasthibandara** [comment]
hi @cyrilletuzi , any reason for closing this?

**Comment by cyrilletuzi** [comment]
First because when `checkJs` mode was released I was very excited to be able to have the power of static types but staying in native JS. But in the end it's really more painful than to be in real TypeScript, especially for that kind of advanced features.

Second, because there was no reaction, so I supposed my early enthusiasm was really not such a good idea.

**Comment by mhegazy** [comment]
> But in the end it's really more painful than to be in real TypeScript, especially for that kind of advanced features.

wondering if you can elaborate about your experience with checkJS..

**Comment by pulasthibandara** [comment]
In my case I'm still a bit reluctant to transpile my `node.js` code or use something like `ts-node`.  Even if its a bit verbose, having the option to define interfaces in .d.ts files and use them in actual js code is really useful for some compile time type checking.

**Comment by cyrilletuzi** [comment]
My main use case for `checkJs` is in Node, but I wouldn't go as far as writing interfaces in jsdoc, it's already quite heavy and messy to just write `@typedef` for objects, and to manage to have the types for all the libraries used.

**Comment by mhegazy** [comment]
> Even if its a bit verbose, having the option to define interfaces in .d.ts files and use them in actual js code is really useful for some compile time type checking.

Why is not `@typedef` not sufficient?

**Comment by mhegazy** [comment]
> My main use case for checkJs is in Node, but I wouldn't go as far as writing interfaces in jsdoc, it's already quite heavy and messy to just write @typedef for objects, and to manage to have the types for all the libraries used.

Have you considered adding .d.ts for your types and using these in your .js files?

**Comment by cyrilletuzi** [comment]
Well, no :/ In my head it was "full JS" or "full TS", but I didn't thought I could mix both worlds. Therefore, `jsdoc` for interfaces is really not needed.

**Comment by pulasthibandara** [comment]
@mhegazy I did try using .d.ts files. But I couldn't figure out how to use the interfaces defined there on the js files.

**Comment by mhegazy** [comment]
>  did try using .d.ts files. But I couldn't figure out how to use the interfaces defined there on the js files.

were both modules?

**Comment by artin-phares** [comment]
Voting for reopening this issue and support '@interface' and '@implements'

@cyrilletuzi 
> First because when checkJs mode was released I was very excited to be able to have the power of static types but staying in native JS. But in the end it's really more painful than to be in real TypeScript, especially for that kind of advanced features.

But if pain comes from lack of supported `jsdoc` features then let's support it, right (why close the issue)?
Or there are any other concerns about `ts` annotations vs `jsdoc` annotations?
